File: desktop/qbot/runtime/recovery_executor.py
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from qbot.persistence.journal import JournalRepository
from qbot.persistence.outbox import OutboxRepository
from qbot.persistence.runs import AgentRunRepository
from qbot.transport import QQTransport

from .recovery import RecoveryAction, RecoveryItem, RecoveryPlan, RecoveryPlanner
from .reply_flow import DurableReplyFlow
from .send import SendExecutor


class RecoveryMode(StrEnum):
    OBSERVE = "observe"
    ASSIST = "assist"


@dataclass(frozen=True, slots=True)
class RecoveryExecution:
    item: RecoveryItem
    outcome: str
    detail: str = ""


@dataclass(frozen=True, slots=True)
class RecoveryReport:
    mode: RecoveryMode
    plan: RecoveryPlan
    executions: tuple[RecoveryExecution, ...]


class RecoveryExecutor:
    """Apply a RecoveryPlan without violating Outbox uncertainty rules."""

    def __init__(
        self,
        *,
        planner: RecoveryPlanner,
        runs: AgentRunRepository,
        outbox: OutboxRepository,
        journal: JournalRepository,
        transport: QQTransport,
        reply_flow: DurableReplyFlow,
    ) -> None:
        self.planner = planner
        self.runs = runs
        self.outbox = outbox
        self.journal = journal
        self.transport = transport
        self.reply_flow = reply_flow
        self.sender = SendExecutor(outbox, transport)

    async def run(self, mode: RecoveryMode | str) -> RecoveryReport:
        resolved_mode = RecoveryMode(mode)
        plan = self.planner.plan()

        if resolved_mode == RecoveryMode.OBSERVE:
            return RecoveryReport(
                mode=resolved_mode,
                plan=plan,
                executions=tuple(
                    RecoveryExecution(
                        item=item,
                        outcome="REPORTED",
                        detail="observe mode does not execute recovery actions",
                    )
                    for item in plan.items
                ),
            )

        executions: list[RecoveryExecution] = []
        for item in plan.items:
            executions.append(await self._execute(item))

        return RecoveryReport(
            mode=resolved_mode,
            plan=plan,
            executions=tuple(executions),
        )

    async def _execute(self, item: RecoveryItem) -> RecoveryExecution:
        if item.action in {
            RecoveryAction.MANUAL_REVIEW,
            RecoveryAction.WAITING_USER,
        }:
            return RecoveryExecution(
                item=item,
                outcome="DEFERRED",
                detail="recovery requires human input",
            )

        if item.action == RecoveryAction.FINALIZE_SENT_RUN:
            run = self.runs.load(item.run_id)
            if run is None:
                return RecoveryExecution(item, "SKIPPED", "AgentRun disappeared")
            if run.status == "EXECUTING":
                self.runs.transition(item.run_id, "SUCCEEDED")
                self._journal_recovery(item, "finalized_sent_run")
                return RecoveryExecution(
                    item,
                    "FINALIZED",
                    "all durable effects were already terminal",
                )
            return RecoveryExecution(
                item,
                "SKIPPED",
                f"AgentRun is no longer EXECUTING: {run.status}",
            )

        if item.action == RecoveryAction.SEND_PENDING:
            if item.outbox_id is None:
                return RecoveryExecution(item, "SKIPPED", "missing outbox_id")
            record = self.outbox.load(item.outbox_id)
            if record.status != "PENDING":
                return RecoveryExecution(
                    item,
                    "SKIPPED",
                    f"Outbox is no longer PENDING: {record.status}",
                )

            result = await self.sender.send(item.outbox_id)
            if result.status == "SENT":
                finalized = self._finalize_if_terminal(item.run_id)
                self._journal_recovery(item, "pending_send_completed")
                return RecoveryExecution(
                    item,
                    "SENT",
                    "pending effect sent"
                    + (" and run finalized" if finalized else ""),
                )
            return RecoveryExecution(
                item,
                result.status,
                result.last_error or "pending send did not reach SENT",
            )

        if item.action == RecoveryAction.RECONCILE_UNKNOWN:
            if item.outbox_id is None:
                return RecoveryExecution(item, "SKIPPED", "missing outbox_id")
            record = self.outbox.load(item.outbox_id)
            if record.status != "SENDING_UNKNOWN":
                return RecoveryExecution(
                    item,
                    "SKIPPED",
                    f"Outbox is no longer SENDING_UNKNOWN: {record.status}",
                )

            result = await self.sender.reconcile(item.outbox_id)
            if result.status == "SENT":
                finalized = self._finalize_if_terminal(item.run_id)
                self._journal_recovery(item, "unknown_send_reconciled")
                return RecoveryExecution(
                    item,
                    "RECONCILED",
                    "delivery confirmed"
                    + (" and run finalized" if finalized else ""),
                )

            return RecoveryExecution(
                item,
                "UNRESOLVED",
                "delivery lookup did not confirm prior delivery; no resend attempted",
            )

        if item.action == RecoveryAction.RESUME_RUN:
            run = self.runs.load(item.run_id)
            if run is None:
                return RecoveryExecution(item, "SKIPPED", "AgentRun disappeared")
            result = await self.reply_flow.execute(
                run_id=run.run_id,
                event_id=run.trigger_event_id,
            )
            self._journal_recovery(item, "run_resumed")
            return RecoveryExecution(
                item,
                result.run_status,
                "run resumed through normal restore/policy/outbox flow",
            )

        return RecoveryExecution(
            item,
            "SKIPPED",
            f"unsupported recovery action: {item.action}",
        )

    def _finalize_if_terminal(self, run_id: str) -> bool:
        run = self.runs.load(run_id)
        if run is None or run.status != "EXECUTING":
            return False

        effects = self.outbox.list_for_run(run_id)
        if not effects:
            return False
        if any(
            effect.status not in {"SENT", "CANCELLED"}
            for effect in effects
        ):
            return False

        self.runs.transition(run_id, "SUCCEEDED")
        return True

    def _journal_recovery(self, item: RecoveryItem, result: str) -> None:
        self.journal.append(
            event_type="RECOVERY_ACTION",
            actor="SYSTEM",
            run_id=item.run_id,
            related_id=item.outbox_id or item.run_id,
            payload={
                "action": item.action.value,
                "result": result,
            },
        )

File: android/app/src/main/java/dev/qbot/android/runtime/RecoveryExecutor.kt
package dev.qbot.android.runtime

import dev.qbot.android.data.AgentRunRepository
import dev.qbot.android.data.OutboxRepository
import dev.qbot.android.transport.OutgoingMessage
import dev.qbot.android.transport.QQTransport

enum class RecoveryMode {
    OBSERVE,
    ASSIST,
}

enum class RecoveryOutcome {
    REPORTED,
    DEFERRED,
    SENT,
    RECONCILED,
    FINALIZED,
    SEND_UNKNOWN,
}

data class RecoveryExecution(
    val item: RecoveryItem,
    val outcome: RecoveryOutcome,
    val detail: String = "",
)

data class RecoveryReport(
    val mode: RecoveryMode,
    val plan: RecoveryPlan,
    val executions: List<RecoveryExecution>,
)

class RecoveryExecutor(
    private val planner: RecoveryPlanner,
    private val runs: AgentRunRepository,
    private val outbox: OutboxRepository,
    private val transport: QQTransport,
    private val sender: SendExecutor = SendExecutor(
        outbox = outbox,
        transport = transport,
    ),
) {
    suspend fun run(
        mode: RecoveryMode = RecoveryMode.ASSIST,
    ): RecoveryReport {
        val plan = planner.plan()
        if (mode == RecoveryMode.OBSERVE) {
            return RecoveryReport(
                mode = mode,
                plan = plan,
                executions = plan.items.map {
                    RecoveryExecution(
                        item = it,
                        outcome = RecoveryOutcome.REPORTED,
                        detail = it.reason,
                    )
                },
            )
        }

        val executions = plan.items.map { item ->
            execute(item)
        }
        return RecoveryReport(
            mode = mode,
            plan = plan,
            executions = executions,
        )
    }

    private suspend fun execute(
        item: RecoveryItem,
    ): RecoveryExecution =
        when (item.action) {
            RecoveryAction.SEND_PENDING ->
                executePending(item)

            RecoveryAction.RECONCILE_UNKNOWN ->
                executeReconcile(item)

            RecoveryAction.FINALIZE_SENT_RUN ->
                executeFinalize(item)

            RecoveryAction.RESUME_RUN ->
                RecoveryExecution(
                    item = item,
                    outcome = RecoveryOutcome.DEFERRED,
                    detail = "decision runtime resume is not enabled in Android v0.5 yet",
                )

            RecoveryAction.MANUAL_REVIEW ->
                RecoveryExecution(
                    item = item,
                    outcome = RecoveryOutcome.DEFERRED,
                    detail = item.reason,
                )

            RecoveryAction.WAITING_USER ->
                RecoveryExecution(
                    item = item,
                    outcome = RecoveryOutcome.DEFERRED,
                    detail = item.reason,
                )
        }

    private suspend fun executePending(
        item: RecoveryItem,
    ): RecoveryExecution {
        val outboxId = requireNotNull(item.outboxId)
        val record = outbox.load(outboxId)
        if (record.status != "PENDING") {
            return RecoveryExecution(
                item = item,
                outcome = RecoveryOutcome.DEFERRED,
                detail = "Outbox is no longer PENDING: " + record.status,
            )
        }

        val message = OutgoingMessage(
            accountId = record.accountId,
            conversationId = record.conversationId,
            dedupeKey = record.dedupeKey,
            text = record.payloadText.orEmpty(),
            replyToMessageId = record.replyToMessageId,
        )
        val readiness = transport.checkSendReadiness(message)
        if (!readiness.ready) {
            return RecoveryExecution(
                item = item,
                outcome = RecoveryOutcome.DEFERRED,
                detail = readiness.reason
                    ?: "transport is not currently ready to send",
            )
        }

        val sent = sender.send(outboxId)
        return when (sent.status) {
            "SENT" -> {
                val finalized = finalizeRunIfEffectsTerminal(item.runId)
                RecoveryExecution(
                    item = item,
                    outcome = RecoveryOutcome.SENT,
                    detail = if (finalized) {
                        "durable PENDING effect sent once; AgentRun finalized locally"
                    } else {
                        "durable PENDING effect sent once"
                    },
                )
            }

            "SENDING_UNKNOWN" -> RecoveryExecution(
                item = item,
                outcome = RecoveryOutcome.SEND_UNKNOWN,
                detail = sent.lastError.orEmpty(),
            )

            else -> RecoveryExecution(
                item = item,
                outcome = RecoveryOutcome.DEFERRED,
                detail = "send completed in state " + sent.status,
            )
        }
    }

    private suspend fun executeReconcile(
        item: RecoveryItem,
    ): RecoveryExecution {
        val outboxId = requireNotNull(item.outboxId)
        val reconciled = sender.reconcile(outboxId)
        return if (reconciled.status == "SENT") {
            val finalized = finalizeRunIfEffectsTerminal(item.runId)
            RecoveryExecution(
                item = item,
                outcome = RecoveryOutcome.RECONCILED,
                detail = if (finalized) {
                    "unknown delivery reconciled without resend; AgentRun finalized locally"
                } else {
                    "unknown delivery reconciled without resend"
                },
            )
        } else {
            RecoveryExecution(
                item = item,
                outcome = RecoveryOutcome.DEFERRED,
                detail = "delivery remains unresolved",
            )
        }
    }

    private suspend fun finalizeRunIfEffectsTerminal(
        runId: String,
    ): Boolean {
        val run = runs.load(runId) ?: return false
        if (run.status != "EXECUTING") {
            return false
        }

        val effects = outbox.listForRun(runId)
        if (
            effects.isEmpty() ||
            effects.any {
                it.status != "SENT" && it.status != "CANCELLED"
            }
        ) {
            return false
        }

        runs.transition(runId, "SUCCEEDED")
        return true
    }

    private suspend fun executeFinalize(
        item: RecoveryItem,
    ): RecoveryExecution {
        val run = runs.load(item.runId)
            ?: return RecoveryExecution(
                item = item,
                outcome = RecoveryOutcome.DEFERRED,
                detail = "AgentRun no longer exists",
            )
        if (run.status != "EXECUTING") {
            return RecoveryExecution(
                item = item,
                outcome = RecoveryOutcome.DEFERRED,
                detail = "AgentRun is no longer EXECUTING: " + run.status,
            )
        }

        runs.transition(item.runId, "SUCCEEDED")
        return RecoveryExecution(
            item = item,
            outcome = RecoveryOutcome.FINALIZED,
            detail = "external effects were already terminal; finalized locally",
        )
    }
}
